filament/src/materials/dof.mat

```diff
@@ -49,17 +44,271 @@ fragment {
+    opacity = cocToAlpha(coc);
+    i += 1.0;
+}
+
+void accumulateForegroundFirst(inout vec4 foreground, inout float opacity, inout float i,
+        highp vec2 pos, float radius, float mip) {
+    float coc = textureLod(materialParams_cocFgBg, pos, mip).r;
+    foreground = textureLod(materialParams_foreground, pos, mip);
+    opacity = cocToAlpha(coc);
+    i += 1.0;
+}
+
+void accumulateBackgroundMirror(inout vec4 background, inout float opacity, inout float i,
+        highp vec2 center, vec2 offset, float radius, float mip) {
+    accumulateBackground(background, opacity, i, center + offset, radius, mip);
+    accumulateBackground(background, opacity, i, center - offset, radius, mip);
+}
+
+void accumulateForegroundMirror(inout vec4 foreground, inout float opacity, inout float i,
+        highp vec2 center, vec2 offset, float radius, float mip) {
+    // The code below is equivalent to:
+    //      accumulateForeground(foreground, opacity, i, center + offset, radius, mip);
+    //      accumulateForeground(foreground, opacity, i, center - offset, radius, mip);
+    // but selects the min coc of opposite samples as a way to guess the color occluded by
+    // the geometry.
+    
+    float coc0 = textureLod(materialParams_cocFgBg, center + offset, mip).r;
+    float coc1 = textureLod(materialParams_cocFgBg, center - offset, mip).r;
+    float coc = min(coc0, coc1);
+    vec4 s0 = textureLod(materialParams_foreground, center + offset, mip);
+    vec4 s1 = textureLod(materialParams_foreground, center - offset, mip);
+    float o = cocToAlpha(coc);
+    float w = intersection(radius, abs(coc), mip) * isForeground(coc);
+    foreground += mix(foreground * rcp(i), s0, w);
+    opacity += mix(opacity * rcp(i), o, w);
+    i += 1.0;
+    foreground += mix(foreground * rcp(i), s1, w);
+    opacity += mix(opacity * rcp(i), o, w);
+    i += 1.0;
+}
+
+float getMipLevel(float kernelSize) {
+    float mip = log2(kernelSize * rcp(RING_COUNT + 0.5));
+    //mip = floor(0.5 + mip);
+    return mip;
+}
+
 void postProcess(inout PostProcessInputs postProcess) {
     highp vec2 uv = variable_vertex.xy;
+    vec2 tiles = textureLod(materialParams_tiles, uv, 0.0).rg;
+
+    /*
+     * Tiles that are neither foreground or background (in focus) can be skipped
+     */
 
-    float depth = textureLod(materialParams_depth, uv, 0.0).r;
-    float coc = getCOC(depth, materialParams.cocParams);
+    if (isTrivialTile(tiles)) {
+        postProcess.color = vec4(0.0);
+        outAlpha = 0.0;
+        return;
+    }
+
+    vec4 foreground = vec4(0.0);
+    vec4 background = vec4(0.0);
+    float fgOpacity = 0.0;
+    float bgOpacity = 0.0;
+
+    // we use the full resolution pixel size because that's the unit the CoC is in
+    highp vec2 fullResPixelSize = 0.5 / vec2(textureSize(materialParams_foreground, 0));
+
+    // When using mipmaping the noise seems to hurt more than it helps
+    //float randomAngle = random(gl_FragCoord.xy) * (2.0 * PI);
```

Comment:
Remove?

Reply:
It's useful to have when testing in matdbg -- since it's not trivial to remember this code, I left it there with a comment.